### PR TITLE
Test nightly wheel build with NumPy 2.0

### PR DESCRIPTION
**Pull request by lagru**
## Description

Concerns #7282. matplotlib pulls in contourpy and contourpy pins numpy<2. So for now, remove matplotlib as a testing dependency and assert that NumPy 2.0 is used during the tests with for the nightly wheels.

I'm currently also piling on a few fixes that we will eventually need to be compatible with NumPy 2.0.

Modules passing locally for NumPy 1 & 2 (skipping some dependencies):
- [x] `skimage._shared`
- [x] `skimage.color`
- [x] `skimage.data`
- [x] `skimage.draw`
- [x] `skimage.exposure`
- [x] `skimage.feature`
- [x] `skimage.filters`
- [x] `skimage.future`
- [x] `skimage.graph`
- [x] `skimage.io`
- [x] `skimage.measure`
- [x] `skimage.metrics`
- [x] `skimage.morphology`
- [x] `skimage.registration`
- [x] `skimage.restoration`
- [x] `skimage.segmentation`
- [x] `skimage.transform`
- [x] `skimage.util`

References:
- [NumPy 2.0 migration guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html)
- [NumPy 2.0-specific advice](https://numpy.org/devdocs/dev/depending_on_numpy.html#numpy-2-0-specific-advice)
- [NEP 50 — Promotion rules for Python scalars](https://numpy.org/neps/nep-0050-scalar-promotion.html#backward-compatibility)

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
...
```

### CODE REVIEW

File: skimage/draw/draw.py [hunk 1]
```diff
@@ -242,6 +242,9 @@ def polygon_perimeter(r, c, shape=None, clip=False):
 
     Examples
     --------
+    >>> import pytest
```
Comment (jarrodmillman) [inline skimage/draw/draw.py:245]: I would prefer using conftest.py to handle skipping tests when matplotlib isn't installed.

Something like this in `conftest.py` should work:
```
try:
    import matplotlib

    has_matplotlib = True
except ImportError:
    has_matplotlib = False

collect_ignore = []

needs_matplotlib = [
    "draw/draw.py",
    "filters/thresholding.py",
    ...
]

if not has_matplotlib:
    collect_ignore += needs_matplotlib
```

Comment (jarrodmillman) [reply skimage/draw/draw.py:245]:
```
needs_matplotlib = [
    "draw/draw.py",
    "filters/thresholding.py",
    ...
]

collect_ignore = []
try:
    import matplotlib

except ImportError:
    collect_ignore += needs_matplotlib
```

Comment (lagru) [reply skimage/draw/draw.py:245]: Yeah, totally with you on that. That has been bothering me some time for cases with pywt and sklearn as well. 

I didn't know about the solution with `conftest.py` so instead I opted to use doctestplus which I've been meaning to propose for some time, see https://github.com/scikit-image/scikit-image/pull/7289. If that PR get's merged I would update the approach taken here. What do you think about going that direction?

File: skimage/util/lookfor.py [hunk 2]
```diff
@@ -1,7 +1,10 @@
-import numpy as np
 import sys
 
+from .._shared.utils import deprecate_func
+from .._vendored.numpy_lookfor import lookfor as _lookfor
```
Comment (lagru) [inline skimage/util/lookfor.py:4]: Highlighting this. Unfortunately we relied on NumPy's lookfor which is promptly removed in NumPy 2.0. Technically, if we don't want to make a breaking release we need to deprecate the function in a deprecation cycle for which we need to vendor NumPy's lookfor.

Though, skimage.lookfor is an interactive function, so we might get away with not vendoring NumPy's lookfor, and just making it return a deprecation warning...

File: skimage/color/colorconv.py [hunk 3]
```diff
@@ -1046,15 +1046,10 @@ def gray2rgba(image, alpha=None, *, channel_axis=-1):
...
     """
-
     arr = np.asarray(image)
-
-    alpha_min, alpha_max = dtype_limits(arr, clip_negative=False)
-
     if alpha is None:
-        alpha = alpha_max
-
-    if not np.can_cast(alpha, arr.dtype):
```
Comment (lagru) [inline skimage/color/colorconv.py:1057]: Highlighting this bit. As of NumPy 2.0, `np.can_cast` no longer supports Python scalars. I think [`np.result_type`](https://numpy.org/devdocs/reference/generated/numpy.result_type.html) still takes values into account when determining the resulting type because `np.min_scalar_type` is used under the hood. So we should be able to get away by comparing if the resulting type is the same as the target type `arr.dtype`. 

@seberg, taking the liberty to Cc you here. :pray: I'm sure I haven't grokked [NEP 50](https://numpy.org/neps/nep-0050-scalar-promotion.html#backward-compatibility) in its entirety yet.

Comment (seberg) [reply skimage/color/colorconv.py:1057]: No, that isn't correct, result types of course does *not* take the value into account.

It ignores the values (same as if it was 0) for Python complex/float/ints.  For out-of-bound *integers* that would mean that the `full` call below should fail if it is out of range, though.  The `result_type` I guess still rejects floats if the array is integer.

Comment (lagru) [reply skimage/color/colorconv.py:1057]: Oh, you seem to be right. I think I was led astray by this bit from the [docstring](https://numpy.org/devdocs/reference/generated/numpy.result_type.html#numpy.result_type)

> Otherwise, [min_scalar_type](https://numpy.org/devdocs/reference/generated/numpy.min_scalar_type.html#numpy.min_scalar_type) is called on each scalar, and the resulting data types are all combined with [promote_types](https://numpy.org/devdocs/reference/generated/numpy.promote_types.html#numpy.promote_types) to produce the return value.

Concering `np.full`, do I misunderstand you? It seems like `np.full` happily overflows out of bounds integers. E.g.
```python
np.full(1, -1, dtype=np.uint8)
# returns array([255], dtype=uint8)
np.full(1, 300, dtype=np.uint8)
# returns array([44], dtype=uint8)
```

Is there a an alternative solution in NumPy for what `np.can_cast` formerly did here? I guess I could just check if the cast is equal to the original value (https://github.com/scikit-image/scikit-image/pull/7288/commits/54a901fa1b954a9d101341117de810ad2bc583e8). I imagine the performance impact will be low as I expect alpha to be a scalar usally.

Comment (lagru) [reply skimage/color/colorconv.py:1057]: Also curious about `np.min_scalar_type(np.iinfo(np.int64).max)` returning `dtype('uint64')`. I would have expected `np.int64` but I guess the promotion path reaches `np.uint64` first?

Comment (seberg) [reply skimage/color/colorconv.py:1057]: arg, you are right, full is weird `np.array(value, dtype=dtype)` would work, or assignment, but full unfortunately does `np.asarray(value)` and *then* assigns.

**Comment by stefanv** [comment]
I wonder if it is worth refactoring `lookfor` in to a package under SP?

File: skimage/util/lookfor.py (hunk 2 above)
Comment (seberg) [reply skimage/util/lookfor.py:4]: If its annoying, I guess we could also vendor it back.  Does anyone still use lookfor?  I feel as an interactive function, it would be a few old-school power users probably (I never used it), thus the feeling we probably can get away with (and no actual program can fail).

Comment (seberg) [reply skimage/util/lookfor.py:4]: Fair, I am not sure if some IDE has search features, which would be nice.  It still seems odd in NumPy to me and probably very rarely used, but if you prefer keeping it for now, I am fine with that too.

Comment (lagru) [reply skimage/util/lookfor.py:4]: I think most IDEs only take the function / method / class name into account when matching for a search string. Or you can do a full search that includes more than just docstrings. @seberg do you have an opinion on how much maintenance work this has been? 

Of all discussed options, I prefer NumPy to just keep it around as that's the most simple option from our side. However, I'd totally get why you might want to remove `lookfor`. In that case I'd vote for the scientific python package option or removing it all-together. I intend vendoring to be temporary solution only.

Comment (seberg) [reply skimage/util/lookfor.py:4]: I am not sure it has been maintanence work, so if there is a PR to restore it I won't mind merging it.
To me it was mostly about namespace clutter as I don't think typical user use it.  IIRC it also returns some nonsense results (i.e. things that are much better in the html docs, but it doesn't find those, while it found docs that belonged just deleted).

File: skimage/morphology/_flood_fill.py [hunk 4]
```diff
@@ -271,16 +272,19 @@ def flood(image, seed_point, *, footprint=None, connectivity=None, tolerance=Non
...
+            min_value, max_value = numeric_dtype_min_max(seed_value.dtype)
+            with np.errstate(over="raise", under="raise"):
+                try:
+                    low_tol = max(min_value, seed_value - tolerance)
+                except (OverflowError, FloatingPointError):
+                    low_tol = min_value
+                try:
+                    high_tol = min(max_value, seed_value + tolerance)
+                except (OverflowError, FloatingPointError):
+                    high_tol = max_value
```
Comment (lagru) [inline skimage/morphology/_flood_fill.py:287]: Some explanation: in NumPy 2.0 we can no longer rely on 
```python
high_tol = min(max_value, seed_value + tolerance)
low_tol = max(min_value, seed_value - tolerance)
```
to correctly bound to the min and max of the dtype. `seed_value` may be a `np.uint8` in which case adding e.g. 379 to it will result in an OverflowError. 

`with np.errstate(over="raise", under="raise")` is necessary because in NumPy 2 cases like
```python
low_tol = max(0, np.uint8(2) - 3)
```

Though, I may be wrong. I find it very tricky to think and test for all edge cases...
now underflow to `np.uint8(255)` because the 3 will be cast to np.uint8 before addtion. In NumPy 1, the right side would result in np.int64(-1).

File: skimage/draw/draw.py (hunk 1 above)
Comment (lagru) [reply skimage/draw/draw.py]: Addressed in 4f65ab74af464b01396490bb330a02dbf7dceed1

File: skimage/io/tests/test_pil.py [hunk 5]
```diff
@@ -165,9 +167,14 @@ def test_jpg_quality_arg():
         assert sim > 0.99
 
 
+@pytest.mark.xfail(
+    Version(np.__version__) >= Version('2.0.0.dev0'),
+    reason='use_plugin("pil") seems broken for now and tifffile is used instead! '
+    'tifffile uses deprecated attribute `ndarray.newbyteorder`',
+)
 def test_imread_uint16_big_endian():
```
Comment (lagru) [inline skimage/io/tests/test_pil.py:175]: Despite this being the test for the PIL plugin, for some reason this triggers an error in tifffile which leads me to believe that the `use_plugin` function may be broken... but that's something for another issue and PR.

Comment (lagru) [reply skimage/io/tests/test_pil.py:175]: See https://github.com/scikit-image/scikit-image/issues/7305.

File: skimage/transform/_geometric.py [hunk 6]
```diff
@@ -281,28 +281,28 @@ class FundamentalMatrixTransform(_GeometricTransform):
...
-           [ 0.0342092683, -0.1135811955,  0.0222823604]])
+    >>> tform_matrix.inverse(dst)  # doctest: +FLOAT_CMP
+    array([[-0.0513591 ,  0.04170974,  0.01213043],
+           [-0.21599496,  0.29193419,  0.00978184],
+           [-0.0079222 ,  0.03758889, -0.00915389],
+           [ 0.14187184, -0.27988959,  0.02476507],
+           [ 0.05890075, -0.07354481, -0.00481342],
+           [-0.21985267,  0.36717464, -0.01482408],
+           [ 0.01339569, -0.03388123,  0.00497605],
+           [ 0.03420927, -0.1135812 ,  0.02228236]])
```
Comment (lagru) [inline skimage/transform/_geometric.py:305]: Because of floating point comparisons these docstrings seem very brittle. I'm not sure what changed with NumPy 2.0. The dtype is still the same. But I think we can relax the precision here somewhat.

Comment (seberg) [reply skimage/transform/_geometric.py:305]: If a current change, it would be using C++ FFT, and maybe that just changed numeric results slightly.

Comment (lagru) [reply skimage/transform/_geometric.py:305]: Hmm, weird. I investigated and I don't think it's that. The *displayed* values seem the same except a lower precision / fewer digits are used. I can reproduce the failure that warranted the updated docstring also on my local main branch with most recent dependencies but our CI seems happy without the fix? Am I missing some local config file that tells NumPy to use a lower precision when displaying values?

Comment (lagru) [reply skimage/transform/_geometric.py:305]: It get's even weirder. If I run `spin test -- --doctest-plusskimage/transform/_geometric.py` the doctest for `FundamentalMatrixTransform` passes. But if I run the complete test suite the example fails because suddenly more significant digits are used to display the results! :thinking:

File: skimage/io/tests/test_tifffile.py [hunk 7]
```diff
@@ -27,7 +27,7 @@ def test_imread_uint16():
 def test_imread_uint16_big_endian():
     expected = np.load(fetch('data/chessboard_GRAY_U8.npy'))
     img = imread(fetch('data/chessboard_GRAY_U16B.tif'))
-    assert img.dtype == np.uint16
+    assert img.dtype == np.dtype("<u2")
```
Comment (lagru) [inline skimage/io/tests/test_tifffile.py:30]: This failed previously because `np.dtype(">u2") == np.uint16` (regardless of endian-ness) will always be false. 

I'm not sure what the original intention of this test was? If it was asserting that big endian is returned, then tifffile seems to have been returning the wrong endian-ness for some time. Otherwise, the test should use `img.dtype.type == np.uint16` to check just the dtype without comparing endians.

**Comment by lagru** [comment]
Can someone tell my why https://github.com/scikit-image/scikit-image/blob/903824fb4f0ed1f52603f22cf6c2c2133a158ae7/skimage/util/dtype.py#L33

is used instead of `np.int32` to build the list of `_supported_types`? It seems like on AMD `np.intc` isn't `np.int32` which leads to [failing tests](https://github.com/scikit-image/scikit-image/actions/runs/7933218915/job/21661489355?pr=7288#step:5:7392) that try to use our convert machinery with `np.int32`. Not sure if this failure is something that is new with NumPy 2 or was something we didn't notice before and was already a problem with NumPy<2.

File: skimage/transform/_geometric.py (hunk 6 above)
Comment (stefanv) [reply skimage/transform/_geometric.py:305]: Nice find!

File: skimage/util/lookfor.py (hunk 2 above)
Comment (stefanv) [reply skimage/util/lookfor.py:4]: Realizing @Carreau may also have thought about this problem.

File: skimage/morphology/_flood_fill.py (hunk 4 above)
Comment (stefanv) [reply skimage/morphology/_flood_fill.py:287]: Yes, but you will get a RuntimeWarning, which should make our test suite error out.

Easiest in this case may be `max(0, int(x) - 3)`.

File: skimage/io/tests/test_tifffile.py (hunk 7 above)
Comment (stefanv) [reply skimage/io/tests/test_tifffile.py:30]: `np.issubdtype`?

Comment (stefanv) [reply skimage/io/tests/test_tifffile.py:30]: `img.dtype.type == np.uint16`, like you suggest, feels nice and clean.

**Comment by stefanv** [comment]
> Can someone tell my why
> 
> https://github.com/scikit-image/scikit-image/blob/903824fb4f0ed1f52603f22cf6c2c2133a158ae7/skimage/util/dtype.py#L33
> 
> is used instead of `np.int32` to build the list of `_supported_types`?

https://github.com/scikit-image/scikit-image/issues/3043#issuecomment-385624092

File: skimage/io/tests/test_tifffile.py (hunk 7 above)
Comment (lagru) [reply skimage/io/tests/test_tifffile.py:30]: I take from your last comment then that the test is about asserting `np.uint16` and that we don't care about the endian-ness. I'll update.

File: skimage/util/lookfor.py (hunk 2 above)
Comment (Carreau) [reply skimage/util/lookfor.py:4]: I don't have any particular solution. 
I plan to have full documentation search with papyri at some point, we can even do some indexing. but it's not ready yet.

**Comment by lagru** [comment]
I don't really follow. In https://github.com/scikit-image/scikit-image/issues/3043#issuecomment-385624092 it is explained that 

> These map directly to the underlying C types, char, short, int, long, long long.
> Every other type is just an alias to one of these (ie, np.t1 is np.t2).

But it seems that this is not true for NumPy 2 on AMD64 and Windows. According to the [tests](https://github.com/scikit-image/scikit-image/actions/runs/7933851980/job/21663558631#step:5:7502) `np.int32` is clearly missing from the `_supported_dtypes`, so none of the supposed underlying C types is an alias of `np.int32`. 

Furthermore, I don't get why the "underlying C types" are of relevance here. `_convert` seems to me as a function that allows converting between dtypes with automatic scaling. So it seems way more explicit to use `np.int8, np.int32, np.int64` etc here like we do with for the floating types.

Additionally, while `np.int64` is covered by `np.longlong`, `_convert(np.linspace(-1, 1))` overflows for the value 1.

File: skimage/util/lookfor.py (hunk 2 above)
Comment (lagru) [reply skimage/util/lookfor.py:4]: Then let's keep it until the status quo changes.

File: skimage/morphology/_flood_fill.py (hunk 4 above)
Comment (lagru) [reply skimage/morphology/_flood_fill.py]: Casting to the appropriate Python scalar seems indeed like the easiest fix. :+1:

File: skimage/_shared/dtype.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        max_ = info.max
+    elif np.issubdtype(dtype, np.dtype(bool)):
+        min_ = dtype.type(False)
+        max_ = dtype.type(True)
+    else:
+        raise ValueError(f"unsupported dtype {dtype!r}")
+    return min_, max_
+
+
+def to_py_scalar(x):
```
Comment (lagru) [inline skimage/_shared/dtype.py:76]: @seberg, I didn't find a solution in the API to cast a NumPy scalar to the corresponding Python scalar (preserving precision). Did I miss an easier solution. If not, should I make a PR for this convenience function to NumPy?

Comment (seberg) [reply skimage/_shared/dtype.py:76]: How is this different from `arr_or_numpy_scalar.item()`?  That one won't convert float128 (or some other things like datetimes), but if you are interested in numbers `.item()` should work fine.

Comment (lagru) [reply skimage/_shared/dtype.py:76]: `.item()` is perfect; I'm only interested in numbers. Thanks that's really useful to be aware of! :blush:

File: doc/source/user_guide/getting_help.rst [hunk 9]
```diff
@@ -26,16 +26,6 @@ Use the ``quick search`` field in the navigation bar of the online
 documentation to find mentions of keywords (segmentation,
 rescaling, denoising, etc.) in the documentation.
 
-API Discovery
```
Comment (stefanv) [inline doc/source/user_guide/getting_help.rst:29]: Why is this taken out, if we still support it?

**Review by jarrodmillman — APPROVED** [review]
LGTM. If we run into any issues, we can fix them in a follow-up PR.

File: skimage/filters/thresholding.py [hunk 10]
```diff
@@ -137,6 +137,9 @@ def try_all_threshold(image, figsize=(8, 5), verbose=True):
 
     Examples
     --------
+    .. testsetup::
```
Comment (stefanv) [inline skimage/filters/thresholding.py:140]: Any way to get these out of the docstrings and into a configuration file, e.g., if we used doctest++?

File: skimage/morphology/tests/test_flood_fill.py [hunk 11]
```diff
@@ -40,7 +40,9 @@ def test_overrange_tolerance_float():
     image *= max_value
 
     expected = np.ones_like(image)
-    output = flood_fill(image, (0, 1), 1.0, tolerance=max_value * 10)
+    with np.errstate(over="ignore"):
+        tolerance = max_value * 10
```
Comment (stefanv) [inline skimage/morphology/tests/test_flood_fill.py:44]: Hrm, is this safe?

File: skimage/transform/_geometric.py [hunk 12]
```diff
@@ -281,28 +281,28 @@ class FundamentalMatrixTransform(_GeometricTransform):
 
     >>> tform_matrix.estimate(src, dst)
     True
-    >>> tform_matrix.params
-    array([[-0.2178588368,  0.4192819131, -0.0343074756],
-           [-0.0717941428,  0.0451643229,  0.0216072614],
-           [ 0.2480621133, -0.4294781423,  0.0221019139]])
+    >>> tform_matrix.params  # doctest: +FLOAT_CMP
```
Comment (stefanv) [inline skimage/transform/_geometric.py:284]: Again, can't this be a global doctest option?

**Review by stefanv — APPROVED** [review]
I think we should merge this and act on any further concerns as they arise. I did make a few comments, but let's discuss those for potential follow-up.

**Comment by stefanv** [comment]
Thank you, @lagru, that was a big job!

File: doc/source/user_guide/getting_help.rst (hunk 9 above)
Comment (lagru) [reply doc/source/user_guide/getting_help.rst:29]: Right now `lookfor` is deprecated and its removal scheduled for 0.26. While that is so, I think it makes sense to remove this section. It's also in the TODO.txt.

To be honest, I forgot that it was marked as deprecated while the discussion in https://github.com/scikit-image/scikit-image/pull/7288#discussion_r1451033431 happened. I don't mind reverting the deprecation and this in a PR if you would like to see it stay...

File: skimage/filters/thresholding.py (hunk 10 above)
Comment (lagru) [reply skimage/filters/thresholding.py:140]: doctest-plus doesn't have an [option](https://github.com/scientific-python/pytest-doctestplus?tab=readme-ov-file#skipping-tests) to do so in a configuration file, I think.

See https://github.com/scikit-image/scikit-image/pull/7289#discussion_r1448907293 for background on the current pattern. Like @soupault I actually like the explicitness of this and didn't like that the other option using `__doctest_skip__` is hiding that doctests are skipped.

Note that these directives don't show up in our rendered HTML docstrings. 

What do you think?

File: skimage/morphology/tests/test_flood_fill.py (hunk 11 above)
Comment (lagru) [reply skimage/morphology/tests/test_flood_fill.py:44]: Good catch, I used the wrong approach to "fix" this as previously the result was using Python scalars while now it results in an `np.float32(inf)` which is not the same as before. I'll address this in a follow-up PR. :+1:

File: skimage/transform/_geometric.py (hunk 12 above)
Comment (lagru) [reply skimage/transform/_geometric.py:284]: There's the [option in `setup.cfg`](https://github.com/scientific-python/pytest-doctestplus?tab=readme-ov-file#floating-point-comparison). Not fond of adding a `setup.cfg` again just for this.

I think the `+FLOAT_CMP` is actually not necessary. I added this before I discovered that another [test](https://github.com/scikit-image/scikit-image/pull/7288#discussion_r1492444288) was changing the precision for floating point comparisons. So I'll remove it in a follow-up PR. 

Though, in cases like this I would still prefer the more explicit local option unless we have a reason to to not use the default precision globally?

Comment (stefanv) [reply skimage/transform/_geometric.py:284]: https://github.com/scientific-python/pytest-doctestplus/issues/185

Comment (stefanv) [reply skimage/transform/_geometric.py:284]: A global config should be sufficient, but fine with having exceptions defined this way.

File: skimage/filters/thresholding.py (hunk 10 above)
Comment (jarrodmillman) [reply skimage/filters/thresholding.py:140]: You could add something like
```
needs_matplotlib = [
    "draw/draw.py",
    "filters/thresholding.py",
    ...
]

collect_ignore = []
try:
    import matplotlib

except ImportError:
    collect_ignore += needs_matplotlib
```
to `skimage/conftest.py`.

Comment (stefanv) [reply skimage/filters/thresholding.py:140]: It would be nice not to have the docstrings littered with statements that make little sense to new readers (for whom these are most useful).

Comment (lagru) [reply skimage/filters/thresholding.py:140]: Hmm, I see the value in that. `.. testsetup` might appear often enough via `help(some_func)` and using IPython's `?` that it's a relevant concern. 

I'd support an alternative solution if it makes it visible that a test is skipped. Otherwise it might get really confusing for contributors. Imagine them not having installed an optional dependency and wondering why the CI is showing red for something they can't reproduce because the doctest is just silently skipped by some `collect_ignore` somewhere they have no idea about.

I can't get the solution suggested by @jarrodmillman to work reliably, it seems like doctests are still collected. Furthermore, it and [`collect_ignore_glob`](https://docs.pytest.org/en/7.1.x/reference/reference.html#globalvar-collect_ignore_glob) can only ignore entire files or directories.

File: skimage/transform/_geometric.py (hunk 12 above)
Comment (pllim) [reply skimage/transform/_geometric.py:284]: For completeness, `saimn` pointed out in https://github.com/scientific-python/pytest-doctestplus/issues/185 that pyproject.toml is supported since https://github.com/scientific-python/pytest-doctestplus/pull/222

File: skimage/filters/thresholding.py (hunk 10 above)
Comment (jarrodmillman) [reply skimage/filters/thresholding.py:140]: We've been using it for NetworkX reliably for several years. Can you describe how it is unreliable?

It does ignore entire files or directories. That hasn't been an issue for NetworkX.

Comment (lagru) [reply skimage/filters/thresholding.py:140]: > Can you describe how it is unreliable?

Trying that out locally just doesn't seem to work. E.g. if I add `"restoration/j_invariant.py",` to https://github.com/scikit-image/scikit-image/blob/9107d523a899b8d3767630941ac28f438bf875d2/skimage/conftest.py#L5-L6

and run
```
spin test -- -v --doctest-plus skimage/restoration/j_invariant.py
```
it's still picked up. I'm guessing it's because I am somehow using the wrong pattern? I couldn't figure out one that works to skip the module.

Also in case of `skimage/restoration/j_invariant.py`  only the doctest of `denoise_invariant` needs to be skipped if `pywt` is not available. Skipping the entire module would skip two doctests unnecessarily. 

If we can make skipping in doctestplus more visible, then I think we can satisfy everyone. I'm currently looking into it in https://github.com/scientific-python/pytest-doctestplus/issues/246...
